### ISSUE TITLE
HtmlToken's Position.Position is off by 1

### ISSUE THREAD

**Issue by cmwoods** [AngleSharp/AngleSharp#787]
### Prerequisites

- [Y] Can you reproduce the problem in a [MWE](https://en.wikipedia.org/wiki/Minimal_working_example)?
- [Y] Are you running the latest version of AngleSharp?
- [Y] Did you check the FAQs to see if that helps you?
- [Y] Are you reporting to the correct repository? (there are multiple AngleSharp libraries, e.g., `AngleSharp.Css` for CSS support)
- [Y] Did you perform a search in the issues?

For more information, see the `CONTRIBUTING` guide.

### Description

The HtmlToken.Position.Position value for every token appears to be off by 1. While line and column numbers are expected to start with an index value of 1, position (aka offset) values are expected to start with an index value of 0.

### Steps to Reproduce

1. Create a memory stream with an initial value/buffer of "<html><body>This is test</body></html>".
2. Create a TextSource instance passing the memory stream as the first parameter.
3. Enumerate the tokens returned from calling Tokenize() extension method on that TextSource.
 
**Expected behavior:** [What you expected to happen]

Position.Position should return the zero based offset corresponding to the TextSource's input stream. For example, for the HTML StartTag token the Position should be Line 1, Col 1, Position 0.

**Actual behavior:** [What actually happened]
For each token, the token's Position.Position is one greater than it should be. For the example, the HTML StartTag token actually has a Position of Line 1, Col 1, Position 1. However the offset in TextSource's Text property that corresponds to the opening '<' for the element/tag is actually at offset 0.

**Environment details:** [OS, .NET Runtime, ...]
Windows, VS2019, .NET Framework 4.6, Any CPU (Pref 64bit).

### Possible Solution
This is likely just a problem with the initial seeding of the Position as the Position.Position for each token is off by 1.

**Comment by FlorianRappl** [comment]
I think this is a matter of terminology - its not "Index" its "Position". Making it 0-based would be a subtle breaking change with a potentially huge impact.

Maybe this should be highlighted in the documentation comment.

Suggestion: The `TextPosition` may have an computed `Index` property yielding `Position - 1`.

**Comment by cmwoods** [comment]
It's not "Index" true but "Position" when there's also line/column information available, to me at least (as I can't rightly speak for others), equates to "stream position" - which is zero based as well (i.e. equivalent to "Index").

I agree that it would be a breaking change. The purist in me would ask for the change but being realistic I am okay, at least for now, with clearly documenting that it is one based instead of zero based. I do think others will run into this, draw the same conclusion ("Position" = "Stream Position"), and not think to consult the documentation (until they notice that it seems off).

Maybe if many others agree with my interpretation then it should be reconsidered for the breaking change.

**Comment by FlorianRappl** [comment]
> Maybe if many others agree with my interpretation then it should be reconsidered for the breaking change.

If many goes to 100% then this is certainly true, however, note that there is a critical implementation detail for the +1 - which is the interpretation of the (consumed) character. In a classical "Index" you have the caret between 2 characters (e.g., 0 meaning between nothing and the first character). In our case we are "standing" at a specific character already, i.e., we have no caret, but rather a selection. The "caret" for the position 1 is thus at 0 (before 1st character) and 1 (after the 1st character) at the same time. We have this because we consumed the character already. This makes the internal mental model much simpler (going forward, but potentially also backward - two-way is the key here) and would introduce many internal breaking changes as well (honestly, these could be circumvented, but one needs to be super careful and one would as a consequence still implement an external breaking change).

The token API was internal anyway and has only been made public due to popular demand. The ones who implemented against it did all use the position as is and did not complain / lived with this behavior (that's not to say it is "common" / "good" / "whatever" - just that there may be other things that are more important than a +1 or not *when* its consistent).

So long story short: I think introducing `Index` and documenting `Position` more carefully should be sufficient.

I'll mark this as enhancement and supply the computed `Index` in the next release.

**Comment by FlorianRappl** [comment]
Landed in `devel`.
